### PR TITLE
NN-5419: add roles to endpoints

### PR DESCRIPTION
**Pull request by neilmendum**
https://dsdmoj.atlassian.net/browse/NN-5419

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftAdjudicationWorkflowController.kt [hunk 1]
```diff
@@ -15,6 +15,7 @@ import uk.gov.justice.digital.hmpps.hmppsmanageadjudicationsapi.config.ErrorResp
 import uk.gov.justice.digital.hmpps.hmppsmanageadjudicationsapi.dtos.ReportedAdjudicationDto
 import uk.gov.justice.digital.hmpps.hmppsmanageadjudicationsapi.services.AdjudicationWorkflowService
 
+@PreAuthorize("hasRole('ADJUDICATIONS_REVIEWER') and hasAuthority('SCOPE_write')")
```
Comment (neilmendum) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftAdjudicationWorkflowController.kt:18]: Where it made sense I moved `@PreAuthorize` from `@Get/Post/PutMapping` and onto the `@RestController`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/MigrateController.kt [hunk 2]
```diff
@@ -52,7 +52,7 @@ data class PunishmentMapping(
   val sanctionSeq: Long,
 )
 
-@PreAuthorize("hasAuthority('SCOPE_write')")
+@PreAuthorize("hasRole('MIGRATE_ADJUDICATIONS') and hasAuthority('SCOPE_write')")
```
Comment (neilmendum) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/MigrateController.kt:55]: For each `SCOPE_write` I thought we'd also want to enforce `MIGRATE_ADJUDICATIONS`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftAdjudicationController.kt [hunk 3]
```diff
@@ -143,6 +143,7 @@ class DraftAdjudicationController(
     ),
   )
   @GetMapping("/my-reports")
+  @PreAuthorize("hasRole('VIEW_ADJUDICATIONS')")
```
Comment (neilmendum) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftAdjudicationController.kt:146]: For each `GET` endpoint without a scope or existing role, I added `VIEW_ADJUDICATIONS`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftAdjudicationController.kt [hunk 4]
```diff
@@ -160,7 +161,7 @@ class DraftAdjudicationController(
     )
 
   @PostMapping
-  @PreAuthorize("hasAuthority('SCOPE_write')")
+  @PreAuthorize("hasRole('ADJUDICATIONS_REVIEWER') and hasAuthority('SCOPE_write')")
```
Comment (neilmendum) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftAdjudicationController.kt:164]: For each `SCOPE_write` (except for migration) I added the role `ADJUDICATIONS_REVIEWER `

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/WitnessesController.kt [hunk 5]
```diff
@@ -13,7 +13,7 @@ import org.springframework.web.bind.annotation.RestController
 import uk.gov.justice.digital.hmpps.hmppsmanageadjudicationsapi.controllers.draft.WitnessesRequest
 import uk.gov.justice.digital.hmpps.hmppsmanageadjudicationsapi.services.reported.WitnessesService
 
-@PreAuthorize("hasAuthority('SCOPE_write')")
+@PreAuthorize("hasRole('ADJUDICATIONS_REVIEWER') and hasAuthority('SCOPE_write')")
```
Comment (timoj58) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/WitnessesController.kt:16]: this is not ALO specific

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/EvidenceController.kt [hunk 6]
```diff
@@ -13,7 +13,7 @@ import org.springframework.web.bind.annotation.RestController
 import uk.gov.justice.digital.hmpps.hmppsmanageadjudicationsapi.controllers.draft.EvidenceRequest
 import uk.gov.justice.digital.hmpps.hmppsmanageadjudicationsapi.services.reported.EvidenceService
 
-@PreAuthorize("hasAuthority('SCOPE_write')")
+@PreAuthorize("hasRole('ADJUDICATIONS_REVIEWER') and hasAuthority('SCOPE_write')")
```
Comment (timoj58) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/EvidenceController.kt:16]: this is not ALO specific

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/DamagesController.kt [hunk 7]
```diff
@@ -13,7 +13,7 @@ import org.springframework.web.bind.annotation.RestController
 import uk.gov.justice.digital.hmpps.hmppsmanageadjudicationsapi.controllers.draft.DamagesRequest
 import uk.gov.justice.digital.hmpps.hmppsmanageadjudicationsapi.services.reported.DamagesService
 
-@PreAuthorize("hasAuthority('SCOPE_write')")
+@PreAuthorize("hasRole('ADJUDICATIONS_REVIEWER') and hasAuthority('SCOPE_write')")
```
Comment (timoj58) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/DamagesController.kt:16]: this is not ALO specific

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/ReportsController.kt [hunk 8]
```diff
@@ -34,6 +34,7 @@ data class AgencyReportCountsDto(
   val transferReviewTotal: Long,
 )
 
+@PreAuthorize("hasRole('ADJUDICATIONS_REVIEWER')")
```
Comment (timoj58) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/ReportsController.kt:37]: some of these reports are not ALO specific.  such as my reports

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftAdjudicationWorkflowController.kt [hunk 9]
```diff
@@ -42,7 +43,6 @@ class DraftAdjudicationWorkflowController(
       ),
     ],
   )
-  @PreAuthorize("hasAuthority('SCOPE_write')")
   @ResponseStatus(HttpStatus.CREATED)
   fun completeDraftAdjudication(@PathVariable(name = "id") id: Long): ReportedAdjudicationDto =
```
Comment (timoj58) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftAdjudicationWorkflowController.kt:47]: this is VIEW role. too

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftOffenceController.kt [hunk 10]
```diff
@@ -43,7 +43,7 @@ class DraftOffenceController(
 
   @PutMapping(value = ["/{id}/offence-details"])
   @Operation(summary = "Set the offence details for the draft adjudication.", description = "At least one set of offence details must be supplied")
-  @PreAuthorize("hasAuthority('SCOPE_write')")
+  @PreAuthorize("hasRole('ADJUDICATIONS_REVIEWER') and hasAuthority('SCOPE_write')")
```
Comment (timoj58) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftOffenceController.kt:46]: this is VIEW too

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/ReportsControllerTest.kt [hunk 11]
```diff
@@ -69,7 +69,7 @@ class ReportsControllerTest : TestControllerBase() {
     }
```
Comment (timoj58) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/ReportsControllerTest.kt:69]: i think most of this controller will be VIEW role too, maybe some specific ALO ones but unlikely

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftAdjudicationController.kt (hunk 4 above)
Comment (neilmendum) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftAdjudicationController.kt]: This was incorrect :) updated with https://github.com/ministryofjustice/hmpps-manage-adjudications-api/pull/467/commits/ec613a63d5ad153abaf7f88ccfe88a77d2fa935f

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/WitnessesController.kt (hunk 5 above)
Comment (neilmendum) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/WitnessesController.kt]: indeed - see https://github.com/ministryofjustice/hmpps-manage-adjudications-api/pull/467#discussion_r1286774976

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/EvidenceController.kt (hunk 6 above)
Comment (neilmendum) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/EvidenceController.kt]: indeed - see https://github.com/ministryofjustice/hmpps-manage-adjudications-api/pull/467#discussion_r1286774976

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/DamagesController.kt (hunk 7 above)
Comment (neilmendum) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/DamagesController.kt]: indeed - see https://github.com/ministryofjustice/hmpps-manage-adjudications-api/pull/467#discussion_r1286774976

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/ReportsController.kt (hunk 8 above)
Comment (neilmendum) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/ReportsController.kt]: indeed - see https://github.com/ministryofjustice/hmpps-manage-adjudications-api/pull/467#discussion_r1286774976

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftAdjudicationWorkflowController.kt (hunk 9 above)
Comment (neilmendum) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftAdjudicationWorkflowController.kt:47]: indeed - see https://github.com/ministryofjustice/hmpps-manage-adjudications-api/pull/467#discussion_r1286774976

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftOffenceController.kt (hunk 10 above)
Comment (neilmendum) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftOffenceController.kt]: indeed - see https://github.com/ministryofjustice/hmpps-manage-adjudications-api/pull/467#discussion_r1286774976

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/ReportsControllerTest.kt (hunk 11 above)
Comment (neilmendum) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/ReportsControllerTest.kt:69]: indeed - see https://github.com/ministryofjustice/hmpps-manage-adjudications-api/pull/467#discussion_r1286774976

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftOffenceController.kt (hunk 10 above)
Comment (neilmendum) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/draft/DraftOffenceController.kt]: indeed - see https://github.com/ministryofjustice/hmpps-manage-adjudications-api/pull/467#discussion_r1286774976

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/ReportedAdjudicationController.kt [hunk 12]
```diff
@@ -52,16 +52,18 @@ class ReportedAdjudicationController(
...
 
   @GetMapping(value = ["/{chargeNumber}/v2"])
+  @PreAuthorize("hasRole('VIEW_ADJUDICATIONS')")
   fun getReportedAdjudicationDetailsV2(@PathVariable(name = "chargeNumber") chargeNumber: String): ReportedAdjudicationResponseV2 =
     reportedAdjudicationService.getReportedAdjudicationDetailsV2(chargeNumber).toResponseV2()
 
   @PutMapping(value = ["/{chargeNumber}/status"])
   @Operation(summary = "Set the status for the reported adjudication.")
-  @PreAuthorize("hasAuthority('SCOPE_write')")
+  @PreAuthorize("hasRole('VIEW_ADJUDICATIONS') and hasAuthority('SCOPE_write')")
```
Comment (timoj58) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/ReportedAdjudicationController.kt:66]: this is an ALO role.  only ALO can set status.  its a bad name for the endpoint, its the review endpoint

Comment (neilmendum) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/ReportedAdjudicationController.kt]: Good spot! updated with https://github.com/ministryofjustice/hmpps-manage-adjudications-api/pull/467/commits/c2bfe5c16690b7beae059cef905d840845f93795

**Review by timoj58 — APPROVED** [review]

**Review by timoj58 — APPROVED** [review]

**Pull request merged**